src/elastio-snap.c

```diff
@@ -3421,7 +3432,7 @@ static MRF_RETURN_TYPE tracing_mrf(struct request_queue *q, struct bio *bio){
 
 	smp_rmb();
 	tracer_for_each(dev, i){	// for each snap device
-		if(!dev || test_bit(UNVERIFIED, &dev->sd_state) || !tracer_queue_matches_bio(dev, bio)) continue;
+		if(!dev || test_bit(UNVERIFIED, &dev->sd_state) || !tracer_matches_bio(dev, bio)) continue;
 
```

Comment:
Hi

[Analysis]

Before 5.8, `tracer_matches_bio` ensures:
1. `queue` is the same (i.e.: the same disk)

But after 5.9, `tracer_matches_bio` only ensures:
1. `fops` is the same (i.e: not 100% the same disk)

Also that `tracer_should_trace_bio` only check the **sector range**,
After 5.9, **we don't know the incoming bio belongs which disk**

[Scenario]
.
After 5.9, we need to make another check for the coming `bio` if it goes into the right **disk**  
before checking `tracer_should_trace_bio` in case the bio COW the wrong disk.


[Issue reproduce]

1. sda1 is in snapshot, sdb1 is not (which means sdb1's bio will also go here)
3. There has a little chance sdb1's bio will pass the check `tracer_should_trace_bio`(same sector range for bio) and further snapshot sdb1's bio on sda1's partition.
4. But we don't want COW the sdb1's bio.

[Conclusion]
1. We only want sdb1's bio find its `orig_mrf`, not go to `tracer_should_trace_bio` if sdb1 is not in snapshot. here we need protect.

Reply:
Hi Nick,

Thanks a lot for the interest and time spend on the analysis of the problem.

You are right, the driver should not perform CoW on the devices which are not tracked.
As I discovered, the LVM devices have shared same `block_device_operations` structure as well as shared request queue. This applies even for the LVM devices in the different volume groups.
In case for the kernels 5.9 and newer, we are replacing `block_device_operations` structure in the tracked devices on the `start_snapshot` IOCTL call. The `tracing_mrf` is the replacement of the original `submit_bio` function in this structure. The new separate struct is allocated for each new tracked device. So, comparison of the fops struct of the bio and base dev for kernels 5.9+ is equivalent to the comparison of the queue in the bio and base dev for kernels 5.8-. We are sure that `continue` will be called for the bio of some other not tracked device.

Regarding the problems of XFS on LVM. They are related to the case, when CoW file's size is exceeded. The problems, presumably, occur due to the delayed flush of the journal (and/or superblock) on XFS and due to the LVM's layered structure, as result, delayed writes and ignored `fsync`. Here is some info https://linux-lvm.redhat.narkive.com/LUYL1j16/fsync-and-lvm The discussion is 13 years old. But the problem is still the same. The issue is captured by the tests `test_transition_fs_sync_cow_full` and `test_transition_mod_sync_cow_full`. And it's not related to the bios flow in the `tracing_mrf` function. Moreover, the problem exists on the kernel versions older than 5.9. There are 2 issues #143 and #145.